Serialization feature support

JiriOndrusek opened the pull request:
fixes  https://github.com/quarkusio/quarkus/issues/14530

GraalVM contains support for serialization https://github.com/oracle/graal/pull/2730

This PR tries to add programmatical approach for extensions. Main idea is to provide generated registration of serialization classes using existing BuildItems. 

**Description**

-  `serialization` field is added into `ReflectiveClassBuildItem `, `ReflectiveHierarchyBuildItem` and `RegisterForReflection` annotation.
- If serialization requirement is detected, generated **AutoFeature** depends on **SerializationFeatue** (there is a "cache" of registered serialized classes, which is required in runtime.
- Each  class for serialization is registered via `com.oracle.svm.reflect.serialize.SerializationSupport.addConstructorAccessor` and reflection is registered via `com.oracle.svm.reflect.serialize.SerializationSupport.addReflections`
- Unfortunately there is a code which handles abstract classes, which had to be rewriten into bytecode generation. This seems to be a weak point of this approach, because it has to be kept in sync with graalVm code.

**Possible optimizations and features for better experience**

- There could be duplicate reflective registrations, one from our buildItem implementation, second from `com.oracle.svm.reflect.serialize.SerializationSupport.addReflections`. This could be avoided.
- There is no need to register classes with `Externalizable` interface. Similar classes could be skipped for serialization registration.
- It would be nice to add some kind of `ReferenceSerializationBuiltItem`, which register all `serializable` references of the target class. It would make registration easier, there will be no requirement to register for example `java.lang.String`, which is typically part of the serialized classes.

~~Problems and TODOs~~

- ~~I hit a problem with several types during its serialization. I'm not sure what is happening. But it has to be investigated and fixed.~~ 
- ~~If `java.util.concurrent.atomic.AtomicBoolean` is registered with this feature, I see following error (registration via config works):~~
> error:org.graalvm.compiler.java.BytecodeParser$BytecodeParserError: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: VarHandle object must be a compile time constant
- ~~If `java.lang.String` is registered with this feature, I see following error (registration via config works):~~
>  error:org.graalvm.compiler.java.BytecodeParser$BytecodeParserError: com.oracle.svm.core.util.VMError$HostedError: Discovered an unresolved calee while parsing java.lang.StringCoding$StringDecoder.decode(StringCoding.java:191)

**Alternative approaches**

It could be feasible to gather all classes registered for serialization, generate serialization-config.xml and attach it to build args. This approach won't duplicate any code from GraalVM.

JUnit test is present.

**Edited** - some problems were solved. See comments.

JiriOndrusek commented:
1. Is this approach acceptable? should I try different way of implementing it?
2. I was not able to find what is causing 
> error:org.graalvm.compiler.java.BytecodeParser$BytecodeParserError: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: VarHandle object must be a compile time constant

in case, that `java.util.concurrent.atomic.AtomicBoolean` is registered via `ReflectiveClassBuildItem.serializationClass(String... className)`. May I as for help?

JiriOndrusek commented:
Both problems (with `AtomicBolean` and `String`) are solved.

From my POV it has to be decided whether we would like to use this path (which contains duplication of code from GraalVM - but in fact it is only one simple if...)

JiriOndrusek commented:
@gsmet, @stuartwdouglas  I'm not sure whether this is desired solution. It works, There are be some things to improve, more tests to add., ...
What should I do, to finish this change (to create PR which could be approved)?

stuartwdouglas commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java:
```diff
@@ -345,6 +361,78 @@ public void write(String s, byte[] bytes) {
+
+                    requiredCatch.returnValue(requiredFeaturesList);
+                }
+                serializationSupport = tc.invokeStaticMethod(
```
There seems to a lot of code here that is generated for each entry, when it should just be run once. We don't want to generate lots of code as there can potentially be lots of entries which can cause problems with the method size limitations. Can you move this code into a function that gets called for each entry rather than duplicating it each time.

JiriOndrusek replied:
Thanks for the comment. I'll look into it and try to fix it.

stuartwdouglas reviewed (changes requested):
Mostly looks good, except for that one comment. 

I also don't really like that we need a JSON file for String but we can look at that later.

JiriOndrusek committed:
Serialization feature support

JiriOndrusek commented:
@stuartwdouglas  I moved almost all code, which was generated for serialization, into a new method and method is called from class registrations (if is required).  

I also added generated code for `Externalizable` classes - to have the same behavior as in GraalVM feature,

stuartwdouglas commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java:
```diff
@@ -53,12 +60,26 @@ public ReflectiveClassBuildItem(boolean constructors, boolean methods, boolean f
+
     private ReflectiveClassBuildItem(boolean constructors, boolean methods, boolean fields, boolean finalFieldsWritable,
             boolean weak, String... className) {
+        this(constructors, methods, fields, false, false, false, className);
```
This is wrong, it needs to pass in the values on finalFieldsWritable and weak, not just false. This is why there are test failures.

JiriOndrusek replied:
Thank you for spotting that. I'm not sure why I changed the original behavior (seems like a copy/paste problem). It is fixed now.

JiriOndrusek commented:
Error in the build does not seem to be linked to the current changes.

>LINK : fatal error LNK1171: unable to load mspdbcore.dll (error code: 1455)

zakkak commented:
>     * If serialization requirement is detected, generated **AutoFeature** depends on **SerializationFeatue** (there is a "cache" of registered serialized classes, which is required in runtime.

@JiriOndrusek thanks for this. Just one question, is it safe to assume that if serialization is not detected this PR doesn't break compatibility with GraalVM 20.3?

JiriOndrusek commented:
@zakkak yes. Only if a reflective class with flag `serilization = true` exists, then SerializationFeature is required

> for (Map.Entry<String, ReflectionInfo> entry : reflectiveClasses.entrySet())
> ...
> if (entry.getValue().serialization) {
   if (registerSerializationMethod == null) {
                    registerSerializationMethod = **createRegisterSerializationForClassMethod(file)**;
                }
                tc.invokeStaticMethod(registerSerializationMethod, clazz);
            }

machi1990 commented on integration-tests/main/src/main/resources/serialization-config.json:
```diff
@@ -0,0 +1,5 @@
+[
+  {
+    "name" : "java.lang.String"
+  }
+]
```
Why is this needed?

JiriOndrusek replied:
This is required for the test. Object used for serialization has a local property String.

> private String name;

gsmet replied:
But isn't the whole point of this feature the ability to not have this sort of file?

machi1990 replied:
> But isn't the whole point of this feature the ability to not have this sort of file?

+1

JiriOndrusek replied:
@gsmet 
For the serialization object, there is an annotation (with serialization flag).
For the String, I'd create ReflectiveClassBuildItem instead of this file (if I would be creating an extension) .  But I'm not sure whether it could be done in the test. - so in extension, this config file won't be needed.

JiriOndrusek replied:
> ReflectiveClassBuildItem.serializationClass("java.lang.String")

JiriOndrusek replied:
Test is aimed to serialize `SomeSerializationObject` I've added String property to be able to compare serialized result.
If there is a way, how to register `java.lang.String` for reflection (with serialization), it would be better approach.

machi1990 replied:
> If there is a way, how to register `java.lang.String` for reflection (with serialization), it would be better approach.

I agree. Otherwise, we can register `String` and other primitive class for serialization automatic.

JiriOndrusek replied:
@machi1990  I'd say that automatic registration of primitive types could be the "next step". I mean that it makes sense to merge this PR (from my POV) and solve primitive types later. This fact is also mentioned in previous comment by @stuartwdouglas 

> I also don't really like that we need a JSON file for String but we can look at that later.

What would be also nice is automatic registration of all dependencies. (like to register serialized object. Local variables and their local variables could be registered automatically/recursively) But I don't know whether this approach is used elsewhere or if it makes sense from the architectural POV.

machi1990 replied:
> @machi1990 I'd say that automatic registration of primitive types could be the "next step". I mean that it makes sense to merge this PR (from my POV) and solve primitive types later. This fact is also mentioned in previous comment by @stuartwdouglas
> 
> > I also don't really like that we need a JSON file for String but we can look at that later.
> 
> What would be also nice is automatic registration of all dependencies. (like to register serialized object. Local variables and their local variables could be registered automatically/recursively) But I don't know whether this approach is used elsewhere or if it makes sense from the architectural POV.

@JiriOndrusek thanks for the reply. Make sense to look at it later, in that case can we create a followup issue for that?. Also, I am sorry for repeating what was already said by others, I totally missed that reply from @stuartwdouglas.

JiriOndrusek replied:
Follow up issue sounds good to me.
@machi1990 I see several ways, how to solve it: 

1. To get rid of registering String class for reflection(serialization) by automatic registration of primitive types
2. or just change test in a way, that config file is not needed (because extensions can register primitives types via `ReflectiveClassBuildItem`)
3. or to try some better approach here?

May be creation of a general issue to propose a way of getting rid of primitives registration makes sense, WDYT?

machi1990 replied:
Thanks @JiriOndrusek. 

I'd say (1), because it is not simply about the test but the user being able to setup their app in a similar way the test is setup so option (2) won't help.  Better to discuss this in the issue that will be created so that (3) can be evoked too.

machi1990 commented on core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java:
```diff
@@ -470,7 +620,8 @@ public void addReflectiveField(Map<String, ReflectionInfo> reflectiveClasses, Re
         if (existing == null) {
-            reflectiveClasses.put(cl, existing = new ReflectionInfo(false, false, false, false, false));
+            reflectiveClasses.put(cl, existing = new ReflectionInfo(false, false, false, false, false,
+                    cl.equals("io.quarkus.it.corestuff.SomeSerializationObject")));
```
This is a test file right? Why is it in the core?

JiriOndrusek replied:
This seems like a something which shouldn't be there. Thanks for catching this. I'll fix it.

JiriOndrusek replied:
fixed, thanks again

gastaldi replied:
Looks like the test file is still referenced here

JiriOndrusek replied:
Sorry for not fixing it correctly. I can not imagine why it is still there, because I remember removing it. I'll fix it again and double check it.

JiriOndrusek replied:
@gastaldi I's fixed, I double checked it. There is no occurrence of `SomeSerializationObject` in  `NativeImageAutoFeatureStep`. Thanks a lot for validating it.

machi1990 reviewed (commented):
Thanks for the  @JiriOndrusek added two questions

JiriOndrusek commented:
Test failures `io.quarkus.oidc.client.OidcClientCredentialsTestCase.testGetTokensOnDemand` don't seem to be linked with current changes. As you can see, tests which are failing now are different then in the report (4 hours ago), even if content of this PR was not changed.

JiriOndrusek commented:
PR is ready to be merged.

ppalaga commented:
Is this good to merge, @gsmet? This blocks a couple of issues in Camel Quarkus.

ppalaga commented:
Is there any hope to get this in Quarkus 2.0.0?

JiriOndrusek commented:
Failed test ` quarkus-integration-test-grpc-streaming` doesn't seem to be linked to this PR.

stuartwdouglas commented:
@gsmet did you have any more concerns about this one?

ppalaga commented:
Giving up hoping to get this in 2.0.0, now hoping for 2.1.0 ;)

geoand commented:
Are we good to go with this one? It looks like it...

stuartwdouglas commented:
If there are any issues we can fix them later, nobody seems to have any more concerns so I don't see any reason to hold it.

JiriOndrusek commented:
There is a small issue/missing feature which would make the use of serialization easier.

Currently all classes used by serialization has to be registered, even the primitive ones. It would be nice to make registration for serialization of some types like String, Boolean, ..... simpler. 
For example: These types could be registered automatically, or there could be some config.property/buildItem to register basic serialization types if required by user.
